weaviate/embedded.py

```diff
@@ -233,6 +240,7 @@ def start(self) -> None:
             )
             self.process = process
         print(f"Started {self.options.binary_path}: process ID {self.process.pid}")
+        os.environ["WEAVIATE_RUNNING_EMBEDDED_PROCESS_ID"] = str(self.process.pid)
```

Comment:
why store it in an env variable and not a member variable?

Reply:
So that `start` can check if another `client` in the same environment has already made a running embedded process outside of the scope of the current `client` allowing to discern between a local and embedded to throw

Reply:
why can't I have two at the same time?

Reply:
I still don't entirely get the point you are making above? What is the problem?

Reply:
The `is_listening()` method only checks that there is an open port

In order for @jfrancoa's `test_connect_to_embedded_with_already_running_local_weaviate` test to not fail, an exception must be raised if a port is listening and is not an embedded instance

For your use-case, an env var will be tracked that contains the port number so that multiple embeddeds can be ran on different ports but none on ports already occupied by another process

Reply:
so when the client crashs/I kill it I cannot start a new embedded afterwards? Because the env var is not cleared?

Why not just check if something is running on the port _before_ we try to start anything. Then we know if it is free or not and can raise an exception in case something else is running

Reply:
That's true, will removing the env var in the `EmbeddedDB.__del__` method suffice?

We want to be able to make two separate clients for the same embedded instance so there needs to be a way to persist state between client scopes within the same environment as evidenced by the `test_embedded_already_running` test. Just because something is running on that port doesn't mean it is an embedded instance so tracking the PID in an env var specific to that port accomplishes that

Reply:
> - connecting to a running local weaviate => exception
> - connecting to a running embedded weaviate => all fine

I don't see a way to do this without tracking the PID of a running embedded instance. Is there another way to tell whether Weaviate has been deployed as a Docker image or as an executable?

Reply:
> > That's true, will removing the env var in the EmbeddedDB.**del** method suffice?
> 
> I am not sure 🤷
> 
> > We want to be able to make two separate clients for the same embedded instance so there needs to be a way to persist state between client scopes within the same environment as evidenced by the test_embedded_already_running test. Just because something is running on that port doesn't mean it is an embedded instance so tracking the PID in an env var specific to that port accomplishes that
> 
> I am sceptical about state tracking and don't really see the difference between:
> 
>     * connecting to a running local weaviate => exception
> 
>     * connecting to a running embedded weaviate => all fine
> 
> 
> why not just make it optional by adding a `connect_to_already_running_weaviate: bool = False` (or whatever name makes sense) to the connect_to_embedded helper?

From my point of view, it does make a difference, in the way that if you run `connect_to_embedded` you expect the client to take care of starting Weaviate as a binary. If other Weaviate instance, running in a different mode than embedded, is blocking from starting up, we should stop it somehow.

Not sure if this will be of use, but when I had to debug this, I checked what process was listening on that port and verifying if the process matched with the embedded binary name. If the binary is running in an embedded mode, it will be executed from the .cache folder within the subfolder weaviate-embedded:

```
lsof -i :8080
COMMAND     PID     USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
weaviate-  8724 jfrancoa   15u  IPv4 0x9d9380b42966c63d      0t0  TCP localhost:http-alt (LISTEN)
Python    76104 jfrancoa    6u  IPv6 0x9d9380b42ef5c965      0t0  TCP localhost:55422->localhost:http-alt (CLOSED)
Python    76104 jfrancoa    8u  IPv6 0x9d9380b42ef5c165      0t0  TCP localhost:55483->localhost:http-alt (CLOSED)

ps aux | grep 8724
jfrancoa          8724   0.4  0.2 409945280  84912   ??  S    Thu09AM  10:26.21 /Users/jfrancoa/.cache/weaviate-embedded/weaviate-v1.23.4-aca78e4081e57c05f18bfb90b48a599baa5fa0917e95625780552142b9a88b9a --host 127.0.0.1 --port 8080 --scheme http
```
That directory is stored in  `DEFAULT_BINARY_PATH = str(Path.home() / ".cache/weaviate-embedded/")`, so if the process listening in that port matches the binary path it means the already running instance is an embedded one, otherwise it will be a local and we can print an error bug saying that a different type of Weaviate instance is already running. Use another port or stop it before starting the embedded one.

Reply:
I see what you mean, but I think we can't get it right in all cases.

For me connect_to_embedded should startup a weaviate instance that is then controlled by the script that is running. If you want to connect to any already running weaviate instance you should use connect_to_local.

Even if we could somehow auto-detect if it is another embedded instance: That instance could shutdown at any moment and that is probably not what most users expect.